### PR TITLE
feat:chatUserColor Var

### PR DESCRIPTION
**Pull request by CKY-**
Description of the Change
added user chat color variable
added associated Chat Api Endpoint

Applicable Issues
https://github.com/crowbartools/Firebot/issues/2415

Testing
used in a chat message command, preset effect and event to verify that the user hex color was returned
returns White instead of string empty to not break CSS or other styling for other use cases.

### CODE REVIEW

**Review by heyaapl — COMMENTED** [review]
Validated that this functions - it successfully retrieved the user color when I tested this.

File: src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                description: "Gets the color for associated user (Ie who triggered command, pressed button, etc)."
+            }
+        ],
+        categories: [VariableCategory.COMMON, VariableCategory.TRIGGER, VariableCategory.USER],
+        possibleDataOutput: [OutputDataType.TEXT]
+    },
+    evaluator: async (trigger, username: string) => {
+        try {
+            let chatColor = "";
+            if (username !== null) {
```
Comment (ebiggz) [inline src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts:29]: Is there a specific need to only check `null` but not `undefined` here (`!== null`)? If not, using `!= null` like we do below should be used

File: src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { ReplaceVariable } from "../../../../../../types/variables";
+import { EffectTrigger } from "../../../../../../shared/effect-constants";
+import { OutputDataType, VariableCategory } from "../../../../../../shared/variable-constants";
+import accountAccess from "../../../../../common/account-access";
+import viewerDatabase from "../../../../../viewers/viewer-database";
+
+const twitchApi = require("../../../../../twitch-api/api");
+
```
Comment (ebiggz) [inline src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts:8]: Let's define a "Default color" variable here so we can reuse it below:
```js
const DEFAULT_COLOR = "#ffffff";
```

File: src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        try {
+            let chatColor = "";
+            if (username !== null) {
+                const viewer = await viewerDatabase.getViewerByUsername(username);
+                if (viewer != null) {
+                    chatColor = await twitchApi.chat.getColorForUser(viewer._id);
+                    if (chatColor == null) {
+                        return "#ffffff";
+                    }
+                    return chatColor;
```
Comment (ebiggz) [inline src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts:36]: Could simplify these lines to:
```js
return chatColor ?? DEFAULT_COLOR;
```

File: src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            } else if (trigger.type === EffectTrigger.EVENT || trigger.type === EffectTrigger.MANUAL) {
+                chatColor = trigger.metadata.eventData.color;
+            }
+            if (chatColor === "") {
+                const streamer = accountAccess.getAccounts().streamer;
+                chatColor = await twitchApi.chat.getColorForUser(streamer.userId);
+                if (chatColor == null) {
+                    return "#ffffff";
+                }
+                return chatColor;
```
Comment (ebiggz) [inline src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts:50]:
```js
return chatColor ?? DEFAULT_COLOR;
```

File: src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            return chatColor;
+        } catch (error) {
+            return "#ffffff";
```
Comment (ebiggz) [inline src/backend/variables/builtin/twitch/chat/message/chat-user-color.ts:54]:
```js
return DEFAULT_COLOR;
```

**Review by ebiggz — CHANGES_REQUESTED** [review]
Nice work! Just some suggestions to simplify & reduce nesting in the variable logic.

**Comment by CKY-** [comment]
@ebiggz thank you for your feedback, I have updated the changes that you requested. I have a feeling that i need to resubmit this  again.. there is something funky going on with my git and the github website where it is trying to pull from master and not V5. when I click the button here to update button.

**Review by ebiggz — APPROVED** [review]
Looks great!

**Pull request merged**